### PR TITLE
feat(connectors): Add invite user to channel option

### PR DESCRIPTION
**Pull request by markfarkas-camunda**
Add "Invite user to channel" method to the Slack connector

camunda/connector-slack#160

### CODE REVIEW

File: src/test/java/io/camunda/connector/slack/model/ConversationsInviteDataTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class ConversationsInviteDataTest {
```
Comment (tmetzke) [inline src/test/java/io/camunda/connector/slack/model/ConversationsInviteDataTest.java:38]: Let's add some more edge cases here, like:
- Channels provided instead of users
- No user provided at all (null input)
- Other data types provided as users (e.g. a number, a list of numbers, a list of mixed Strings and numbers)

**Review by tmetzke — CHANGES_REQUESTED** [review]
Really good overall, I like it, well done 👍 
I added some minor suggestions here and there, but nothing major. Mostly about code style or test cases we should consider additionally.

File: element-templates/slack-connector.json [hunk 2]
```diff
@@ -169,6 +177,45 @@
...
+        }
+      },
+      "condition": {
+        "property": "method",
+        "equals": "conversations.invite"
+      }
+    },
+    {
+      "label": "Users",
+      "description": "A single user, or comma separated list of users, or a valid list of strings as a FEEL expression",
```
[the remark this answers is not in the document]
Comment (markfarkas-camunda) [reply element-templates/slack-connector.json:203]: Wouldn't this imply that only usernames can be provided, although we support email and id? Also would it be clear for the user that he can also provide only one single user?

**Comment by markfarkas-camunda** [comment]
Here are some screenshots about the connector template in the modeler:

Slack connector template 1
Slack connector template 2
Slack connector template 3

File: element-templates/slack-connector.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmetzke) [reply element-templates/slack-connector.json:203]: I think we can expect users to understand that a "list of users" implies that you can also only add one user, yes.
We can probably add an email and an ID here as well in the examples, sure. 🙂

**Review by bastiankoerber — APPROVED** [review]
template looks good to me 👍

**Comment by markfarkas-camunda** [comment]
@tmetzke I've just pushed my fixes. Could you please take a look at it?

File: src/main/java/io/camunda/connector/slack/model/ChatPostMessageData.java [hunk 3]
```diff
@@ -27,7 +30,14 @@ public class ChatPostMessageData implements SlackRequestData {
...
     if (channel.startsWith("@")) {
-      channel = DataLookupService.getUserIdByName(channel.substring(1), methodsClient);
+      List<String> userIds =
+          DataLookupService.getIdListByUserNameList(
+              Arrays.asList(channel.substring(1)), methodsClient);
+      channel =
+          Optional.ofNullable(userIds)
+              .filter(list -> !list.isEmpty())
+              .map(list -> list.get(0))
+              .orElseThrow(() -> new RuntimeException("Unable to find users by name"));
```
Comment (tmetzke) [inline src/main/java/io/camunda/connector/slack/model/ChatPostMessageData.java:40]: I think it's fine to keep a specific method for this in the DataLookupService. It is the concern of that lookup service to handle that it reuses another method. The code here should ideally stay as-is. Plus, the exception message we had before should also ideally not change.

File: src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java [hunk 4]
```diff
@@ -32,9 +32,14 @@ public SlackResponse invoke(MethodsClient methodsClient) throws SlackApiExceptio
     List<String> userInputAsList = null;
     if (users instanceof Collection<?>) {
       userInputAsList = (ArrayList<String>) users;
```
Comment (tmetzke) [inline src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java:34]: Will this here throw an exception if I provide a list of numbers?

File: src/test/java/io/camunda/connector/slack/SlackFunctionTest.java [hunk 5]
```diff
@@ -226,7 +227,7 @@ void execute_shouldThrowExceptionWhenUserListResponseIsFail(String input) {
             RuntimeException.class,
             () -> slackFunction.execute(context),
             "RuntimeException was expected");
-    assertThat(thrown.getMessage()).contains("Unable to find user with name: JohnDou");
+    assertThat(thrown.getMessage()).contains("Unable to find users by name");
```
Comment (tmetzke) [inline src/test/java/io/camunda/connector/slack/SlackFunctionTest.java:230]: Ideally, we tell the user which name couldn't be resolved if we have that information at hand.

File: src/test/java/io/camunda/connector/slack/SlackFunctionTest.java [hunk 6]
```diff
@@ -226,7 +227,7 @@ void execute_shouldThrowExceptionWhenUserListResponseIsFail(String input) {
             RuntimeException.class,
```
Comment (tmetzke) [inline src/test/java/io/camunda/connector/slack/SlackFunctionTest.java:227]: We can try to be as specific as possible with the exception we receive. You can also use `catchThrowable` instead from AssertJ. Then you can do `assertThat(thrown).isInstanceOf(XYZException.class).hasMessageContaining(...)`

File: src/test/java/io/camunda/connector/slack/model/ConversationsInviteDataTest.java [hunk 7]
```diff
@@ -65,6 +65,54 @@ void invoke_shouldThrowExceptionWhenUserNotFoundByEmail() throws SlackApiExcepti
...
+        assertThrows(
+            RuntimeException.class,
+            () -> conversationsInviteData.invoke(methodsClient),
+            "RuntimeException was expected");
+    assertThat(thrown.getMessage())
+        .contains("Invalid input type for users. Supported types are: List<String> and String");
+  }
+
+  @Test
+  void invoke_shouldThrowExceptionWhenNullInputForUsers() {
```
Comment (tmetzke) [inline src/test/java/io/camunda/connector/slack/model/ConversationsInviteDataTest.java:101]: This will probably be handled by the validation of the input data when we execute the function. Maybe it's better to test the interactions on the `execute` level of the Slack function rather than on the level of the input object itself.

File: src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java (hunk 4 above)
Comment (tmetzke) [reply src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java:34]: Let's create test cases for this as well.

Comment (markfarkas-camunda) [reply src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java:34]: It throws an exception, but not our exception. 
It throws: class java.util.ImmutableCollections$List12 cannot be cast to class java.util.ArrayList (java.util.ImmutableCollections$List12 and java.util.ArrayList are in module java.base of loader 'bootstrap')

Comment (markfarkas-camunda) [reply src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java:34]: @tmetzke it will throw this error:
class java.lang.Integer cannot be cast to class java.lang.String (java.lang.Integer and java.lang.String are in module java.base of loader 'bootstrap')
Is this acceptable?

Comment (tmetzke) [reply src/main/java/io/camunda/connector/slack/model/ConversationsInviteData.java:34]: Two things:
- We should be as generic as possible here with the type of the user list. We could maybe cast to `Collection<String>` or at least `List<String>` instead of `ArrayList<String>`. If we test for `if (users instanceof Collection)`, we can only cast it to a collection, not a list, actually.
- Let's maybe go for something generic here that is then handled in `DataLookupService#getUserIdsFromNameOrEmail`. For example, we could simply use `Collection<?>` as the type of the `userInput` (renamed from `userInputAsList`). This can be passed on to `DataLookupService#getUserIdsFromUsers` (renamed from `getUserIdsFromNameOrEmail`). In that method, you go through the user elements and check each one's type. For those not being a String, you can either throw an error or silently ignore them.

With the proposed way, we avoid crude class cast exceptions and can take in a lot more types of data. What do you think? I can also commit a proposal, if you think that is easier. 🙂

**Comment by markfarkas-camunda** [comment]
@tmetzke could you please review my latest commit?

**Review by tmetzke — APPROVED** [review]
Awesome job!
